server.py

```diff
@@ -1,36 +1,376 @@
+                data = json.load(f)
+            logger.info("Loading in coaches DB from file: %s", file)
+            self._db = {c["id"]: Coach.from_json(c) for c in data["data"]}
+            self._next_id = data["next_id"]
+        except FileNotFoundError:
+            logger.info("Coaches DB file not found, creating now: %s", file)
+            self.persist(file)
+
+
+# The in-memory database of coaches.
+coach_db: CoachDB
+
+# The file used to persist the coach DB.
+COACH_DB_FILE = "./coach_db.json"
+
+
+# ------------------------------------------------------------------------------
+# Helpers
+# ------------------------------------------------------------------------------
+
+
+def _parse_languages(languages: str) -> Dict[str, int]:
+    ret = dict()
+    if not languages:
+        return ret
+    for lang in languages.split(","):
+        lang_name, proficiency = lang.split(":")
+        ret[lang_name] = int(proficiency)
+    return ret
+
+
+# ------------------------------------------------------------------------------
+# Route handlers
+# ------------------------------------------------------------------------------
+
+
+def get_coaches() -> List[Coach]:
+    return list(coach_db)
+
+
+def get_coach(coach_id: int) -> Coach:
+    return coach_db.get(coach_id)
+
+
+def create_coach(data: CoachJSON) -> Coach:
+    coach = Coach(**data)
+    db_coach = coach_db.add(coach)
+    return db_coach
+
+
+def edit_coach(coach_id: int, data: CoachJSON) -> Coach:
+    kwargs = get_coach(coach_id).to_json()
+    kwargs.update(data)
+    updated_coach = Coach(**kwargs)
+    db_coach = coach_db.update_entry(updated_coach)
+    return db_coach
+
+
+def delete_coach(coach_id: int) -> None:
+    coach_db.remove(coach_id)
+
+
+def get_coach_matches(data: CoachJSON) -> List[Tuple[int, Coach]]:
+    """
+    Get coaches based on how well they match the given data.
+
+    :param data:
+        Data to match on.
+    :return:
+        A list of tuples containing the match score and the matched coach.
+    """
+
+    def get_lang_match_score(
+        c: Coach, languages: Dict[str, int]
+    ) -> Optional[Tuple[str, int]]:
+        match_langs = [L for L in languages if L in c.languages]
+        return max(
+            (0, *[30 - 2 * (languages[L] * c.languages[L]) for L in match_langs])
+        )
+
+    coach_matches = []
+    for coach in get_coaches():
+        if not coach.available:
+            continue
+        match_score = 0
+        match_score += get_lang_match_score(coach, data.get("languages", []))
+        if data.get("need") in coach.need:
+            match_score += 20
+        if data.get("rights") in coach.rights:
+            match_score += 10
+        if data.get("housing") in coach.housing:
+            match_score += 10
+        if data.get("gender") == coach.gender:
+            match_score += 5
+        if "birth_year" in data:
+            match_score += max(0, 10 - abs(data["birth_year"] - coach.birth_year))
```

Comment:
I'd probably make this 5 not 10, since age doesn't matter as much

Reply:
yeah reasonable, perhaps `max(0, 5 - abs(data["birth_year"] - coach.birth_year)//2)` so that 5 years apart does better than 10 years apart?